clusterctl/clusterdeployer/clusterdeployer.go

```diff
@@ -3,36 +3,313 @@ package clusterdeployer
+	GetKubeConfig(master *clusterv1.Machine) (string, error)
+}
+
 // Can provision a kubernetes cluster
 type ClusterProvisioner interface {
 	Create() error
 	Delete() error
 	GetKubeconfig() (string, error)
 }
 
+// Provides interaction with a cluster
+type ClusterClient interface {
+	Apply(string) error
+	WaitForClusterV1alpha1Ready() error
+	GetClusterObjects() ([]*clusterv1.Cluster, error)
+	GetMachineObjects() ([]*clusterv1.Machine, error)
+	CreateClusterObject(*clusterv1.Cluster) error
+	CreateMachineObjects([]*clusterv1.Machine) error
+	UpdateClusterObjectEndpoint(string) error
+	Close() error
+}
+
+// Can create cluster clients
+type ClusterClientFactory interface {
+	ClusterClient(string) (ClusterClient, error)
+}
+
 type ClusterDeployer struct {
 	externalProvisioner    ClusterProvisioner
+	clientFactory          ClusterClientFactory
+	provider               ProviderDeployer
+	providerComponents     string
+	kubeconfigOutput       string
 	cleanupExternalCluster bool
 }
 
-func New(externalProvisioner ClusterProvisioner, cleanupExternalCluster bool) *ClusterDeployer {
+func New(
+	externalProvisioner ClusterProvisioner,
+	clientFactory ClusterClientFactory,
+	provider ProviderDeployer,
+	providerComponents string,
+	kubeconfigOutput string,
+	cleanupExternalCluster bool) *ClusterDeployer {
 	return &ClusterDeployer{
 		externalProvisioner:    externalProvisioner,
+		clientFactory:          clientFactory,
+		provider:               provider,
+		providerComponents:     providerComponents,
+		kubeconfigOutput:       kubeconfigOutput,
 		cleanupExternalCluster: cleanupExternalCluster,
 	}
 }
 
 // Creates the a cluster from the provided cluster definition and machine list.
-func (d *ClusterDeployer) Create(_ *clusterv1.Cluster, _ []*clusterv1.Machine) error {
+func (d *ClusterDeployer) Create(cluster *clusterv1.Cluster, machines []*clusterv1.Machine) error {
+	master, nodes, err := splitMachineRoles(machines)
+	if err != nil {
+		return fmt.Errorf("unable to seperate master machines from node machines: %v", err)
+	}
+
+	glog.V(1).Info("Spinning up external cluster.")
 	if err := d.externalProvisioner.Create(); err != nil {
 		return fmt.Errorf("could not create external control plane: %v", err)
 	}
 	if d.cleanupExternalCluster {
 		defer d.externalProvisioner.Delete()
 	}
-	return errors.NotImplementedError // Not fully functional yet.
+
+	glog.V(1).Info("Applying Cluster API stack to external cluster.")
+	externalKubeconfig, err := d.externalProvisioner.GetKubeconfig()
+	if err != nil {
+		return fmt.Errorf("unable to get external cluster kubeconfig: %v", err)
+	}
+	externalClient, err := d.clientFactory.ClusterClient(externalKubeconfig)
+	if err != nil {
+		return fmt.Errorf("unable to create external client:%v", err)
+	}
+	defer func() {
+		err := externalClient.Close()
+		if err != nil {
+			glog.Error("Could not close external client")
+		}
+	}()
+
+	glog.V(2).Info("Applying Cluster API APIServer to external cluster.")
```

Comment:
while the code in this method is relatively straightforward, there is a lot to keep track of.
does it make sense to add helper methods for creating objects with the external client one for the internal client?

Reply:
Broke it out as much as I could while preserving flow.